// xtoolkit/src/util/index.js
/**
 * Created by exolution on 17/1/4.
 */
const ChildProcess = require('child_process');
const Fs = require('fs');
const Path = require('path');
exports.npmInstall = function (name, cwd, tnpm) {
    let args = ['install', name, '--loglevel=error'];
    if (tnpm) {
        args.push('--registry=http://registry.npm.taobao.org');
    }
    let result = ChildProcess.spawnSync(process.platform === 'win32' ? 'npm.cmd' : 'npm', args, {
        stdio: ['inherit', 'ignore', 'inherit'],
        cwd: cwd
    })

};
exports.homePath = function () {
    let home=process.env[(process.platform.slice(0, 3) == 'win') ? 'USERPROFILE' : 'HOME'];
    return Path.join(home, '.xtoolkit');
};

exports.modulePath=function(){
    return Path.join(this.homePath(), 'node_modules');
}
exports.init = function () {
    let nodeModulePath=this.modulePath();
    if (!Fs.existsSync(this.homePath())){
        Fs.mkdirSync(this.homePath());
        Fs.mkdirSync(nodeModulePath);
    }
    else if(!Fs.existsSync(nodeModulePath)){
        Fs.mkdirSync(nodeModulePath);
    }
    return nodeModulePath;
};
